File: packages/web-client/src/services/authentication.ts
import HttpClient from '../class/http';
import User from '../class/user';
import GlobalOptions from '../class/global_options';
import { bus, tokenReceivedEvent } from '../class/event-bus'

import {
    Identity, LoginOptions, LoginResponse, ValidateCodeResponse, VerifyTokenResponse, BaseResponse
} from '../types/auth';


class AuthService {

    private static instance: AuthService;
    private http: HttpClient;
    private token?: string | null;

    get isLogin() {
        return !!this.token;
    }

    private constructor() {
        this.http = new HttpClient({ baseUrl: GlobalOptions.host });
    }

    public static getInstance(): AuthService {

        if (AuthService.instance == null) {
            AuthService.instance = new AuthService();
        }

        return AuthService.instance;
    }

    private emitToken() {

        bus.publish(tokenReceivedEvent({ token: this.token || '' }))

    }

    private saveSession() {
        if (this.token)
            localStorage.setItem('token', this.token)
    }

    /**
     * Login with last session if you pass allowSave=true in last login.
     * 
     * @return user 
     */
    loginWithLastSession() {

        // load token
        this.token = localStorage.getItem('token');

        this.validateToken(this.token || '')
            .then(({ user }: { user: any }) => new User({
                email: user.email,
                phone: user.phone,
                id: user.id,
                permission: user['permission']
            }))
    }

    /**
     * Login as an anonymous user and get a token.
     */
    loginAsAnonymous() {
        return this.http.get<LoginResponse>('/user/loginAnonymous')
            .then(body => {
                this.token = body.token
                this.emitToken()
                return body;
            })
    }

    /**
     * Login az an registered user.
     * 
     * @param options
     * @param options.idType the type of user identity
     * @param options.id user identity
     * @param options.password user password
     */
    login(options: LoginOptions, allowSave: boolean) {
        return this.http.post<LoginResponse>('/user/login', options)
            .then(body => {
                this.token = body.token
                this.emitToken()

                if (allowSave) this.saveSession();
            })
            // verify token and get user object
            .then(_ => {
                return this.validateToken(this.token || '')
                    .then(({ user }: { user: any }) => new User({
                        email: user.email,
                        phone: user.phone,
                        id: user.id,
                        permission: user['permission']
                    }))
            })
    }

    validateToken(token: string) {
        return this.http.post<VerifyTokenResponse>('verify/token', { token });
    }

    /**
     * Register user identity, 
     * first step for creating new account
     * 
     * @param options
     * @param options.idType the type of user identity
     * @param options.id user identity
     */
    registerIdentity(identity: Identity) {
        return this.http.post<BaseResponse>('/user/register_id', identity)
    }

    /**
     * Send verification code to server,
     * second step for creating new account.
     * 
     * @param code verification code  
     */
    validateCode(options: { code: string, id: string }) {
        return this.http.post<ValidateCodeResponse>('/user/validateCode', options);
    }

    /**
     * Submit password,
     * third step for creating new account.
     * 
     * @param options 
     * @param options.id user identity
     * @param options.password user password
     * @param options.code verification code
     */
    submitPassword(options: { id: string, password: string, code: string }) {
        return this.http.post<BaseResponse>('/user/submit_password', options)
    }

    /**
     * Change password.
     * 
     * @param options 
     * @param options.id user identity
     * @param options.password user password
     * @param options.code verification code
     */
    changePassword(options: { id: string, password: string, code: string }) {
        return this.http.post<BaseResponse>('/user/change_password', options)
    }
}


export default AuthService;

File: packages/flutter/lib/services/authentication.dart
import 'package:flutter/foundation.dart';
import 'package:mrest_flutter/src/class/user.dart';

import '../src/class/http_client.dart';
import '../src/class/global_options.dart';
import '../src/class/response.dart';
import '../src/class/event_bus.dart';

enum IDType { email, phone }

class AuthService {
  final _eventBus = EventBus();
  final HttpClient _http = HttpClient();
  String _token;
  bool isLogin = false;

  String get baseUrl => GlobalOptions().baseUrl;

  /*
   * Singleton pattern
   */
  static final AuthService _instance = AuthService._internal();
  AuthService._internal() {
    _eventBus.tokenStream.listen((token) {
      isLogin = true;
    });
  }

  factory AuthService() {
    return _instance;
  }
  //-- End Singleton pattern

  // Emit token to other service listeners.
  void _emitToken() {
    _eventBus.tokenSink.add(_token);
  }

  Future<void> _waiteUntilTokenEmitted() {
    return Future.doWhile(() async {
      if (isLogin)
        return false;
      else {
        await Future.delayed(Duration(milliseconds: 500));
        return true;
      }
    });
  }

  /// Login as an anonymouse user
  Future<LoginResponse> loginAsAnonymous() {
    return _http.get(baseUrl + '/user/loginAnonymous').then((body) {
      Map bodyMap = body as Map;
      _token = bodyMap['token'];
      _emitToken();
      return LoginResponse(false, null, bodyMap['status'], bodyMap['token']);
    }).then((value) async {
      await _waiteUntilTokenEmitted();
      return value;
    }).catchError((value) {
      return LoginResponse(true, value, 'fail', null);
    });
  }

  /// login as an registered user
  Future<User> login({
    @required IDType idType,
    @required String id,
    @required String password,
  }) {
    Map data = {
      'idType': (idType.index == IDType.email.index) ? 'email' : 'phone',
      'id': id,
      'password': password,
    };

    return _http.post(baseUrl + '/user/login', body: data).then((body) {
      Map bodyMap = body as Map;
      _token = bodyMap['token'];
      _emitToken();
      return validateToken(_token);
    }).then((value) async {
      await _waiteUntilTokenEmitted();
      return value;
    }).catchError((value) {
      throw LoginResponse(true, value, 'fail', null);
    });
  }

  Future<User> validateToken(String token) {
    Map data = {
      'token': token,
    };

    return _http.post(baseUrl + '/verify/token', body: data).then((body) {
      Map userData = body['user'];
      return User(
        id: userData['id'],
        type: userData['type'],
        email: userData['email'],
        phone: userData['phone'],
        permission: userData['permission'],
      );
    }).catchError((error) {
      BaseResponse(true, error, 'fail');
    });
  }

  /// Register user identity,
  /// first step for creating new account
  Future<BaseResponse> registerIndentity({
    @required IDType idType,
    @required String id,
  }) {
    Map data = {
      'idType': (idType.index == IDType.email.index) ? 'email' : 'phone',
      'id': id,
    };

    return _http.post(baseUrl + '/user/register_id', body: data).then((value) {
      Map body = value as Map;
      return BaseResponse(false, null, body['status']);
    }).catchError((value) => throw BaseResponse(true, value, 'fail'));
  }

  /// Send verification code to server,
  /// second step for creating new account.
  Future<ValidateCodeResponse> validateCode(String id, String code) {
    Map data = {
      'id': id,
      'code': code,
    };

    return _http.post(baseUrl + '/user/validateCode', body: data).then((value) {
      Map body = value as Map;
      return ValidateCodeResponse(false, null, body['status'], body['isValid']);
    }).catchError(
        (value) => throw ValidateCodeResponse(true, value, 'fail', false));
  }

  /// Submit password,
  /// third step for creating new account.
  Future<BaseResponse> submitPassword(String id, String code, String password) {
    Map data = {
      'id': id,
      'code': code,
      'password': password,
    };

    return _http
        .post(baseUrl + '/user/submit_password', body: data)
        .then((value) {
      Map body = value as Map;
      return BaseResponse(false, null, body['status']);
    }).catchError((value) => throw BaseResponse(true, value, 'fail'));
  }

  /// Change password
  Future<BaseResponse> changePassword(String id, String code, String password) {
    Map data = {
      'id': id,
      'code': code,
      'password': password,
    };

    return _http
        .post(baseUrl + '/user/change_password', body: data)
        .then((value) {
      Map body = value as Map;
      return BaseResponse(false, null, body['status']);
    }).catchError((value) => throw BaseResponse(true, value, 'fail'));
  }
}
